### PR TITLE
Seth Verrill

### PR DESCRIPTION
**Pull request by sethverrill**
I'm happy with everything through iteration 3. I got a start on iteration 4, but would need a bit more time to complete it.

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -29,6 +29,7 @@ GEM
 PLATFORMS
   arm64-darwin-20
   arm64-darwin-21
+  arm64-darwin-23
```
Comment (nicktu12) [inline Gemfile.lock:32]: Huh? Why did you need to add this?

File: lib/facility.rb [hunk 2]
```diff
@@ -1,14 +1,82 @@
...
+      @registered_vehicles << vehicle
+      @collected_fees += calculate_fee(vehicle)
+      vehicle
+    else
+      nil
+    end
+  end
+
+  def determine_plate_type(vehicle)
+    if vehicle.year <= 1999
```
Comment (nicktu12) [inline lib/facility.rb:32]: Good control flow here

File: lib/facility.rb [hunk 3]
```diff
@@ -1,14 +1,82 @@
...
+      :antique
+    elsif vehicle.engine == :ev
+      :ev
+    else
+      :regular
+    end
+  end
+
+  def calculate_fee(vehicle)
```
Comment (nicktu12) [inline lib/facility.rb:41]: Love that you broke this functionality out into its own method - SRP!

File: lib/facility.rb [hunk 4]
```diff
@@ -1,14 +1,82 @@
...
+      25
+    elsif vehicle.plate_type == :ev
+      200
+    else
+      100
+    end
+  end
+
+  def administer_written_test(registrant)
+    if @services.include?('Written Test') &&
```
Comment (nicktu12) [inline lib/facility.rb:52]: Great chaining of conditions in this conditional

File: lib/facility_factory.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  def create_facilities(source, state)
+    facilities = []
+    source.each do |data|
+      facilities << create_facility(data, state)
+    end
+    facilities
+  end
+
+  def create_facility(data, state)
+    case state
```
Comment (nicktu12) [inline lib/facility_factory.rb:15]: switch statement 🔥

File: lib/facility_factory.rb [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      data[:state],
+      data[:zipcode],
+    ].compact.join(', ')
+
+    facility = Facility.new({
+      name: data[:name],
+      address: address,
+      phone: data[:phone],
+      holidays_closed: data[:holidaysclosed]
+    })
```
Comment (nicktu12) [inline lib/facility_factory.rb:78]: These three methods are concise and readable -nice!

File: lib/registrant.rb [hunk 7]
```diff
@@ -0,0 +1,56 @@
+class Registrant
+  attr_reader :name,
+              :age,
+              :permit,
```
Comment (nicktu12) [inline lib/registrant.rb:4]: No need for permit here, as this data is exposed via permit?

File: lib/registrant.rb [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    @license_data[:license]
+  end
+
+  def renew_license
+    if has_license?
+      @license_data[:renewed] = true  
+      true    
+    else
+      false
+    end
```
Comment (nicktu12) [inline lib/registrant.rb:54]: This class is great - many small methods promoting SRP

File: lib/stats.rb [hunk 9]
```diff
@@ -0,0 +1,57 @@
+require './spec/spec_helper'
+
+puts "STATISTICS!!!"
```
Comment (nicktu12) [inline lib/stats.rb:3]: ANALYTICS!!!!! Nice extension seth

File: lib/stats.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+  county_count = Hash.new(0)
+  wa_evs_data.each do |county|
+    county_count[county[:county]] += 1
+    
+  end    
+    highest_county = county_count.max_by {|_, count| count}
+    "The county with the highest number of EVs is #{highest_county[0]} with #{highest_county[1]} registrations!"
+end
+
```
Comment (nicktu12) [inline lib/stats.rb:49]: Extra whitespace here - reduce the number of lines in this file!

File: lib/vehicle_factory.rb [hunk 11]
```diff
@@ -0,0 +1,34 @@
+class VehicleFactory
+
+  def initialize
+  end
+
+
+  def create_vehicles(source)
+    vehicles = []
+    source.each do |data|
+    vehicles << create_vehicle(data)
```
Comment (nicktu12) [inline lib/vehicle_factory.rb:10]: watch your indentation here

File: spec/facility_factory_spec.rb [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  end
+
+  it 'can create New York facilities' do
+    ny_data = @dds.ny_dmv_office_locations
+    facilities = @factory.create_facilities(ny_data, 'NY')
+
+    expect(facilities).to all(be_an_instance_of(Facility))
+    expect(facilities.size).to eq(ny_data.size)
+    
+    twenty_first_facility = facilities[20]
```
Comment (nicktu12) [inline spec/facility_factory_spec.rb:37]: You can probably just test the first element in the array for these tests, but I like the ingenuity.

File: spec/facility_spec.rb [hunk 13]
```diff
@@ -21,6 +29,96 @@
...
+      @facility.register_vehicle(@cruz)
+      expect(@facility.collected_fees).to eq (100)
+      @facility.register_vehicle(@camaro)
+      expect(@facility.collected_fees).to eq (125)
+      @facility.register_vehicle(@bolt)
+      expect(@facility.collected_fees).to eq (325)
+    end
+
+    it 'will not perform a service it does not provide' do
+      expect(@facility.register_vehicle(@cruz)).to eq(nil)
```
Comment (nicktu12) [inline spec/facility_spec.rb:60]: Good sad path test here

File: spec/facility_spec.rb [hunk 14]
```diff
@@ -21,6 +29,96 @@
...
+
+  end
+    
+  describe '#can provide license services' do
+    it 'gives written test when registrant is eligible'do
+      @facility.add_service('Written Test')  
+      @registrant_1.earn_permit   
+
+      expect(@facility.administer_written_test(@registrant_1)).to be true
+      expect(@registrant_1.passed_written_test?).to be true
```
Comment (nicktu12) [inline spec/facility_spec.rb:86]: Testing side effects! Nice

File: spec/registrant_spec.rb [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+      expect(@registrant_2.name).to eq("Penny")
+    end
+
+    it 'has an age' do
+
+      expect(@registrant_1.age).to eq(18)
+      expect(@registrant_2.age).to eq(15)
+    end
+
+    it 'has a permit' do
```
Comment (nicktu12) [inline spec/registrant_spec.rb:28]: Could probably test all the initialize functionality in fewer tests, meaning name, age, and permit could all be tested together

File: spec/registrant_spec.rb [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+      expect(@registrant_2.permit?).to eq false
+    end
+
+    it 'can earn a permit' do
+      expect(@registrant_2.permit?).to eq false
+      @registrant_2.earn_permit
+      expect(@registrant_2.permit?).to eq true
+    end
+
+    it 'cannot lose permit once true' do
```
Comment (nicktu12) [inline spec/registrant_spec.rb:40]: Edge case testing here!! Nice

File: spec/vehicle_factory_spec.rb [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+    @factory = VehicleFactory.new
+    @dds = DmvDataService.new
+  end
+
+  it 'exists' do
+    expect(@factory).to be_an_instance_of(VehicleFactory)
+  end
+
+  it 'works with fake vehicles' do
+    cars = [
```
Comment (nicktu12) [inline spec/vehicle_factory_spec.rb:14]: Love that you used a smaller dataset here - this data can live in its own file next time.

File: spec/vehicle_factory_spec.rb [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+      {vin: '123456789abcdefgh', year: 2012, make: 'Chevrolet', model: 'Cruz', engine: :ice},
+      {vin: '987654321abcdefgh', year: 2019, make: 'Chevrolet', model: 'Bolt', engine: :ev},
+      {vin: '1a2b3c4d5e6f', year: 1969, make: 'Chevrolet', model: 'Camaro', engine: :ice}
+    ]
+    vehicles = @factory.create_vehicles(cars)
+    first_vehicle = vehicles[0]
```
Comment (nicktu12) [inline spec/vehicle_factory_spec.rb:20]: Variable assignments might be unnecessary here - just test vehicle[0].vin, vehicle[1].year, etc

File: Gemfile.lock (hunk 1 above)
Comment (sethverrill) [reply Gemfile.lock:32]: When I uninstalled and reinstalled ruby my Gemfiles changes. I just wanted to keep it consistent with what was on my machine.
